plugin/storage/badger/spanstore/cache.go

```diff
@@ -90,67 +102,99 @@ func (c *CacheStore) loadOperations(service string) {
 			keyTTL := it.Item().ExpiresAt()
 			if _, found := c.operations[service]; !found {
-				c.operations[service] = make(map[string]uint64)
+				c.operations[service] = make(map[model.SpanKind]map[string]uint64)
 			}
-
-			if v, found := c.operations[service][operationName]; found {
+			var kind model.SpanKind
+			err := it.Item().Value(func(val []byte) error {
+				if val == nil {
+					kind = model.SpanKindUnspecified
+					return nil
+				}
+				var err error
+				kind, err = model.SpanKindFromString(string(val))
+				return err
+			})
+			if err != nil {
+				kind = model.SpanKindUnspecified
+			}
+			if _, found := c.operations[service][kind]; !found {
+				c.operations[service][kind] = make(map[string]uint64)
+			}
+			if v, found := c.operations[service][kind][operationName]; found {
 				if v > keyTTL {
 					continue
 				}
 			}
-			c.operations[service][operationName] = keyTTL
+			c.operations[service][kind][operationName] = keyTTL
 		}
 		return nil
 	})
 }
 
 // Update caches the results of service and service + operation indexes and maintains their TTL
-func (c *CacheStore) Update(service, operation string, expireTime uint64) {
+func (c *CacheStore) Update(service, operation string, kind model.SpanKind, expireTime uint64) {
 	c.cacheLock.Lock()
-
 	c.services[service] = expireTime
 	if _, ok := c.operations[service]; !ok {
-		c.operations[service] = make(map[string]uint64)
+		c.operations[service] = make(map[model.SpanKind]map[string]uint64)
+	}
+	if _, ok := c.operations[service][kind]; !ok {
+		c.operations[service][kind] = make(map[string]uint64)
 	}
-	c.operations[service][operation] = expireTime
+	c.operations[service][kind][operation] = expireTime
 	c.cacheLock.Unlock()
 }
 
 // GetOperations returns all operations for a specific service & spanKind traced by Jaeger
-func (c *CacheStore) GetOperations(service string) ([]spanstore.Operation, error) {
-	operations := make([]string, 0, len(c.services))
+func (c *CacheStore) GetOperations(service string, kind string) ([]spanstore.Operation, error) {
+	operations := make([]spanstore.Operation, 0, len(c.services))
 	//nolint: gosec // G115
 	t := uint64(time.Now().Unix())
 	c.cacheLock.Lock()
 	defer c.cacheLock.Unlock()
-
 	if v, ok := c.services[service]; ok {
 		if v < t {
 			// Expired, remove
 			delete(c.services, service)
 			delete(c.operations, service)
 			return []spanstore.Operation{}, nil // empty slice rather than nil
 		}
-		for o, e := range c.operations[service] {
-			if e > t {
-				operations = append(operations, o)
-			} else {
-				delete(c.operations[service], o)
+		if kind != "" {
+			spanKind, err := model.SpanKindFromString(kind)
+			if err != nil {
+				return nil, err
+			}
+			for o, e := range c.operations[service][spanKind] {
+				operations = insertOperations(c, operations, service, o, spanKind, t, e)
+				sort.Slice(operations, func(i, j int) bool {
+					return operations[i].Name < operations[j].Name
+				})
+			}
```

Comment:
I don't understand why this part is needed. In the `else` section, you could do
```
for sKind := range c.operations[service] {
  if kind != "" && kind != sKind { continue }
```
to skip unwanted kinds, while the rest of the code is the same whether you're filtering by kind or not.

Reply:
But why to iterate all the kinds when kind is already provided in the query? The currrent approach is: When kind is provided in the query, then directly jump to that service and kind but when it's empty only then iterate all kinds.

Reply:
because it makes the code more complex. Iterating through a collection that has at most 2-3 entries is trivial performance-wise.